src/basic/chattr-util.c

```diff
@@ -10,59 +10,81 @@
+                if (fd < 0)
+                        return -errno;
+        }
 
         if (fstat(fd, &st) < 0)
                 return -errno;
 
-        /* Explicitly check whether this is a regular file or
-         * directory. If it is anything else (such as a device node or
-         * fifo), then the ioctl will not hit the file systems but
-         * possibly drivers, where the ioctl might have different
-         * effects. Notably, DRM is using the same ioctl() number. */
+        /* Explicitly check whether this is a regular file or directory. If it is anything else (such
+         * as a device node or fifo), then the ioctl will not hit the file systems but possibly
+         * drivers, where the ioctl might have different effects. Notably, DRM is using the same
+         * ioctl() number. */
 
         if (!S_ISDIR(st.st_mode) && !S_ISREG(st.st_mode))
                 return -ENOTTY;
 
-        if (mask == 0 && !previous)
+        if (mask == 0 && !ret_previous && !ret_final)
                 return 0;
 
         if (ioctl(fd, FS_IOC_GETFLAGS, &old_attr) < 0)
                 return -errno;
 
         new_attr = (old_attr & ~mask) | (value & mask);
         if (new_attr == old_attr) {
-                if (previous)
-                        *previous = old_attr;
+                if (ret_previous)
+                        *ret_previous = old_attr;
+                if (ret_final)
+                        *ret_final = old_attr;
                 return 0;
         }
 
-        if (ioctl(fd, FS_IOC_SETFLAGS, &new_attr) < 0)
+        if (ioctl(fd, FS_IOC_SETFLAGS, &new_attr) >= 0) {
+                if (ret_previous)
+                        *ret_previous = old_attr;
+                if (ret_final)
+                        *ret_final = new_attr;
+                return 1;
+        }
+
+        if (errno != EINVAL || !fallback)
                 return -errno;
 
-        if (previous)
-                *previous = old_attr;
+        /* When -EINVAL is returned, we assume that incompatible attributes are simultaneously
+         * specified. E.g., compress(c) and nocow(C) attributes cannot be set to files on btrfs.
+         * As a fallback, let's try to set attributes one by one. */
 
-        return 1;
-}
+        unsigned current_attr = old_attr;
+        for (unsigned i = 0; i < sizeof(unsigned) * 8; i++) {
+                unsigned new_one, mask_one = 1u << i;
 
-int chattr_path(const char *p, unsigned value, unsigned mask, unsigned *previous) {
-        _cleanup_close_ int fd = -1;
+                if (!FLAGS_SET(mask, mask_one))
+                        continue;
 
-        assert(p);
+                new_one = UPDATE_FLAG(current_attr, mask_one, FLAGS_SET(value, mask_one));
+                if (new_one == current_attr)
+                        continue;
 
-        if (mask == 0)
-                return 0;
+                if (ioctl(fd, FS_IOC_SETFLAGS, &new_one) >= 0)
+                        current_attr = new_one;
```

Comment:
maybe even safer would be to read the new flags mask back via FS_IOC_GETFLAGS again instead of assuming it was applied 1:1, maybe some weird fs does some internal mangling, we might never know.